### ISSUE TITLE
Support 2024.12

### ISSUE THREAD

**Issue by ev-br** [data-apis/array-api-compat#223]
This is a tracking issue for the 2024.12 revision support in array-api-compat.

The 2024.12 is in a draft still, so this list below is preliminary:

- [x] `diff` : https://github.com/data-apis/array-api-compat/pull/224
- [x] `nextafter` : available in unwrapped namespaces
- [x] `cumulative_prod` : https://github.com/data-apis/array-api-compat/pull/224
- [x] `reciprocal` : available in unwrapped namespaces
- [x] `count_nonzero` : https://github.com/data-apis/array-api-compat/pull/224
- [x] `take_along_axis` : https://github.com/data-apis/array-api-compat/pull/224
- [x] `max_dimensions` in `capabilities` : https://github.com/data-apis/array-api-compat/pull/224
- [x] `mean` is defined for complex arrays : defined in unwrapped namespaces
- [x] `result_type` accepts python scalars : https://github.com/data-apis/array-api-compat/pull/224
.
- [ ] Device-dependent type promotion rules, per https://github.com/data-apis/array-api/pull/889

**Comment by crusaderky** [comment]
take_along_axis in dask is a very complicated matter when there are multiple chunks: https://github.com/dask/dask/issues/3663

**Comment by mdhaber** [comment]
Can we start implementing the draft standard in array API compat, especially `result_type`?

Before scalars were allowed in `array_namespace` and `result type`, I don't think it would be unusual do something like:
```python3
xp = array_namespace(x, y)
x = xp.asarray(x)
y = xp.asarray(y)
dtype = result_type(x, y)
```

especially since the following woudl not be acceptable according to the standard if one of `x` or `y` can be a Python scalar:

```python3
xp = array_namespace(x, y)
dtype = result_type(x, y)  # fails for array_api_strict if `x` or `y` is a Python scalar
```

But if the intent of `result_type` is to find the correct result dtype of an operation between `x` and `y`, the former can produce an incorrect result when scalars are involved, e.g.:
```python3
from array_api_compat import array_namespace, numpy as np
x = np.asarray([1, 2, 3], dtype=np.float32)
y = 3.

xp = array_namespace(x, y)
x_ = np.asarray(x)  # float32
y_ = xp.asarray(y)  # float64

xp.result_type(x_, y_)  # float64
(x*y).dtype  # float32
``` 

So currently, we are between a rock and hard place: the code we might naturally write now (and code that was OK before scalars would be accepted) is not really correct now that scalars are involved.

**Comment by ev-br** [comment]
Yes, the work is ongoing, help appreciated :-). The aim is indeed to have `-compat` and `-strict` releases to support 2024.12 soon after the spec is finalized.

The current state is:


- `array-api-strict`: draft implementation is at https://github.com/data-apis/array-api-strict/pull/119 

<details>

```
In [12]: import array_api_strict as xp

In [13]: xp.result_type(xp.ones(3, dtype=xp.float32), 3)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[13], line 1
----> 1 xp.result_type(xp.ones(3, dtype=xp.float32), 3)

File ~/repos/array-api-strict/array_api_strict/_data_type_functions.py:238, in result_type(*arrays_and_dtypes)
    235     return result
    237 if get_array_api_strict_flags()['api_version'] <= '2023.12':
--> 238     raise TypeError("result_type() inputs must be array_api arrays or dtypes")
    240 # promote python scalars given the result_type for all arrays/dtypes
    241 from ._creation_functions import empty

TypeError: result_type() inputs must be array_api arrays or dtypes

In [14]: xp.set_array_api_strict_flags(api_version='2024.12')
/home/br/repos/array-api-strict/array_api_strict/_flags.py:144: UserWarning: The 2024.12 version of the array API specification is in draft status. Not all features are implemented in array_api_strict, some functions may not be fully tested, and behaviors are subject to change before the final standard release.
  warnings.warn(f"The {draft_version} version of the array API specification is in draft status. Not all features are implemented in array_api_strict, some functions may not be fully tested, and behaviors are subject to change before the final standard release.")

In [15]: xp.result_type(xp.ones(3, dtype=xp.float32), 3)
Out[15]: array_api_strict.float32

In [16]: xp.result_type(xp.ones(3, dtype=xp.uint8), 3)
Out[16]: array_api_strict.uint8
```

</details>


- `array-api-compat`: this is WIP at https://github.com/data-apis/array-api-compat/pull/224

- `array-api-tests`: not much of 2024.12 is tested, so help here would be most welcome. Especially so if you like working with `hypothesis` :-).

**Comment by mdhaber** [comment]
Great. I was tempted to update `result_type` but wasn't sure if it was fair game yet. I'll post here if I start to work on it (but wouldn't want that to discourage anyone else in the meantime).

**Comment by ev-br** [comment]
I'd say it's totally fair game. We won't merge PRs to -strict and -compat until a matching spec PR lands, that's pretty much the only boundary condition.

**Comment by ev-br** [comment]
The bulk of 2024.12 support landed in version 1.11, released today. I'm sure there will be multiple improvements; let's track them separately anyway.
